File: connect4_c/connect4_engine.c
#include "connect4_engine.h"

int place_token(int player, int column, int num_rows,
                int num_columns, int board[num_rows][num_columns]) {
    int column_index = num_columns - 1;

    if ((column <= column_index) && (column >= 0)) {
        int open_row = open_space(column, num_rows, num_columns, board);
        if (open_row > -1) {
            board[open_row][column] = player;
            return player; // This player was successfully placed
        }
    }
    return -1; // The value should still be empty
}

int winner(int num_rows, int num_columns, int length_to_win,
           int array[num_rows][num_columns]) {
    /* There are four ways to win, but only one way will
     * be greater than -1 each time if there is a win. In this way,
     * we can keep track of the wins with an array of size 4. */
    int wins[4] = {-1};
    int player_one_win = -1;
    int player_two_win = -1;
    int board_full = 0;

    wins[0] = check_winner_vertical(
            num_rows, num_columns, length_to_win, array);
    wins[1] = check_winner_horizontal(
            num_rows, num_columns, length_to_win, array);
    wins[2] = check_winner_diagonal_right_up(
            num_rows, num_columns, length_to_win, array);
    wins[3] = check_winner_diagonal_left_up(
            num_rows, num_columns, length_to_win, array);

    board_full = check_full_board(num_rows, num_columns, array);

    for (int i = 0; i < 4; i++) {
        if (wins[i] == 0)
          player_one_win = 1;
        if (wins[i] == 1)
          player_two_win = 1;
    }

    if (player_one_win == 1) {
        return 0;
    } else if (player_two_win == 1) {
        return 1;
    }

    if (board_full == num_columns) {
        return 2;
    }
    return -1;
}

int check_winner_vertical(int num_rows, int num_columns,
                          int length_to_win, int array[num_rows][num_columns]) {
    int win_count = 0;
    int index_player = -1;

    for (int col = 0; col < num_columns - 1; col++) {
        for (int row = num_rows - 1; row > 0; --row) {
            int current_index = array[row][col];
            if ((win_count == 0) && (current_index != -1)) {
                ++win_count;
            }
            if ((current_index == array[row - 1][col]) && (current_index != -1)) {
                ++win_count;
                index_player = array[row][col];
            } else {
                win_count = 0;
                index_player = -1;
            }
            if (win_count == length_to_win) {
                return index_player;
            }
        }
    }
    return -1;
}

int check_winner_horizontal(int num_rows,
                            int num_columns,
                            int length_to_win,
                            int array[num_rows][num_columns]) {
    int win_count = 0;
    int index_player = -1;

    for (int row = 0; row < num_rows; row++) {
        for (int col = 0; col < num_columns - 1; col++) {
            int current_index = array[row][col];

            if ((win_count == 0) && (current_index != -1)) {
                ++win_count;
            }
            if ((current_index == array[row][col + 1]) && (current_index != -1)) {
                ++win_count;
                index_player = array[row][col];
            } else {
                win_count = 0;
                index_player = -1;
            }
            if (win_count == length_to_win) {
                return index_player;
            }
        }
    }
    return -1;
}

int check_winner_diagonal_right_up(int num_rows,
                                   int num_columns,
                                   int length_to_win,
                                   int board[num_rows][num_columns]) {
    int win_count = -1;
    int extra = 0;
    // The increment for how many "diagonals away from" the left we are
    for (int diag_left = 2; diag_left < num_rows * 2 - 3; diag_left++) {
        extra = 0;
        if (diag_left < num_rows) {
            extra = 0;
        } else {
            extra = diag_left - num_rows + 1;
        }
        win_count = -1;
        for (int index = extra; index <= diag_left - extra; index++) {
            int row = index;
            int col = diag_left - index;

            if ((win_count == 0) && (board[row][col] != -1)) {
                ++win_count;
            }
            if ((board[row - 1][col + 1] == board[row][col])
                && (board[row][col] != -1)) {
                ++win_count;
            } else {
                win_count = 0;
            }

            if (win_count >= length_to_win) {
                return board[row][col];
            }
        }
    }
    return -1;
}

int check_winner_diagonal_left_up(int num_rows,
                                  int num_columns,
                                  int length_to_win,
                                  int board[num_rows][num_columns]) {
    int win_count = -1;
    int extra = 0;
    // The increment for how many "diagonals away from" the left we are
    // We can start with 3 diagonals in, as no board will be under
    // three pieces in length, nor less than three to win.
    for (int diag_right = 2; diag_right < num_rows * 2 - 3; diag_right++) {
        extra = 0;
        if (diag_right < num_rows) {
            extra = 0;
        } else {
            // "Over peak" amount of extra counts
            // Since we keep choosing a new index even though the size
            // of the diagonals is now decreasing
            extra = diag_right - num_rows + 1;
        }
        win_count = -1;
        for (int index = extra; index <= diag_right - extra; index++) {
            int row = index;
            int col = (num_rows - 1) - (diag_right - index);
            if ((win_count == 0) && (board[row][col] != -1)) {
                ++win_count;
            }
            if ((board[row - 1][col - 1] == board[row][col]) &&
                (board[row][col] != -1)) {
                ++win_count;
            } else {
                win_count = 0;
            }
            if (win_count >= length_to_win) {
                return board[row][col];
            }
        }
    }
    return -1;
}

int open_space(int column, int num_rows, int num_columns,
               int board[num_rows][num_columns]) {
    int current_index;

    for (int row = num_rows - 1; row > -1; --row) {
        current_index = board[row][column];

        if (current_index < 0) {
            return row;
        }
    }
    return -1; // No open spaces were found
}

int check_full_board(int num_rows, int num_columns,
                     int board[num_rows][num_columns]) {
    int full_count = 0; // The number of columns that are full
    for (int col = 0; col < num_columns; col++) {
        if (board[0][col] != -1) {
            full_count++;
        }
    }
    return full_count;
}

File: project/connect4_python/Connect4Engine.py
#!/usr/bin/env python

from Board import Board


class Connect4Engine:
    def __init__(self, rows, columns, length_to_win):
        self.board = Board(rows, columns)
        self.length_to_win = length_to_win

    def place_token(self, player, column):
        if self.board.columns > column >= 0:
            open_row = self.open_space(column)
            if open_row > -1:
                self.board.grid[open_row][column] = player
                return player  # This player was successfully placed
        return -1  # The value should still be empty

    def winner(self):
        wins = []
        player_one_win = -1
        player_two_win = -1
        wins.append(self.__check_winner_vertical())
        wins.append(self.__check_winner_horizontal())
        wins.append(self.__check_winner_diagonal_up_right())
        wins.append(self.__check_winner_diagonal_left_down())
        board_full = self.__check_full_board()
        for win in wins:
            if win == 0:
                player_one_win = 1
            if win == 1:
                player_two_win = 1
        if player_one_win == 1:
            return 0
        elif player_two_win == 1:
            return 1
        if board_full == self.board.columns:
            return 2
        return -1

    def __check_winner_vertical(self):
        win_count = 0
        for column in xrange(0, self.board.columns - 1):
            for row in xrange(0, self.board.rows - 1):
                current_index = self.board.grid[row][column]
                if win_count == 0 and current_index != -1:
                    win_count += 1
                if current_index == self.board.grid[row - 1][column] and current_index != -1:
                    win_count += 1
                    index_player = self.board.grid[row][column]
                else:
                    win_count = 0
                    index_player = -1
                if win_count == self.length_to_win:
                    return index_player
        return -1

    def __check_winner_horizontal(self):
        win_count = 0
        for row in xrange(0, self.board.rows - 1):
            for column in xrange(0, self.board.columns - 1):
                current_index = self.board.grid[row][column]
                if win_count == 0 and current_index != -1:
                    win_count += 1
                if current_index == self.board.grid[row][column + 1] and current_index != -1:
                    win_count += 1
                    index_player = self.board.grid[row][column]
                else:
                    win_count = 0
                    index_player = -1
                if win_count == self.length_to_win:
                    return index_player
        return -1

    def __check_winner_diagonal_left_down(self):
        rows = self.board.rows
        columns = self.board.columns
        for d_slice in xrange(rows + columns - 1):
            slice_elements = []
            skip_end = 0 if d_slice < columns else d_slice - columns + 1
            skip_start = 0 if d_slice < rows else d_slice - rows + 1
            for j in xrange(d_slice - skip_start, skip_end + 1, -1):
                slice_elements.append(
                    self.board.grid[j][d_slice - j]
                )
            win_count = 0
            for i in xrange(1, len(slice_elements) - 1):
                if win_count == 0 and slice_elements[i] > -1:
                    win_count += 1
                if slice_elements[i + 1] == slice_elements[i] and slice_elements[i] != -1:
                    win_count += 1
                else:
                    win_count = 0
                if win_count >= self.length_to_win:
                    return slice_elements[i]
        return -1

    def __check_winner_diagonal_up_right(self):
        rows = self.board.rows
        columns = self.board.columns
        for d_slice in xrange(rows + columns - 1):
            slice_elements = []
            skip_end = 0 if d_slice < columns else d_slice - columns + 1
            skip_start = 0 if d_slice < rows else d_slice - rows + 1
            for j in xrange(d_slice - skip_start, skip_end + 1, -1):
                slice_elements.append(
                    self.board.grid[rows - j - 1][d_slice - j]
                )
            win_count = 0
            for i in xrange(0, len(slice_elements) - 1):
                if win_count == 0 and slice_elements[i] > -1:
                    win_count += 1
                if slice_elements[i + 1] == slice_elements[i] and slice_elements[i] != -1:
                    win_count += 1
                else:
                    win_count = 0
                if win_count >= self.length_to_win:
                    return slice_elements[i]
        return -1

    def open_space(self, column):
        for row in xrange(self.board.rows - 1, -1, -1):
            current_index = self.board.grid[row][column]
            if current_index < 0:
                return row
        return -1  # No open spaces were found

    def __check_full_board(self):
        full_count = 0  # The number of columns that are full
        for column in xrange(0, self.board.columns):
            if self.board.grid[0][column] != -1:
                full_count += 1
        return full_count
